### PR TITLE
uncrustify: add auto uncrustify with blacklist

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
Here is a proposal for the discussion in #4363. 
- This PR adds a tool in dist/tools that applies the current uncrustify-riot.cfg config to all tracked files that don't match an entry in a blacklist.txt file.
- It also includes a `check.sh` that could be used from CI. Checks if files in a branch have been uncrustified (close to a automatic style check).
- Last but not least, provides a blacklist.txt with core and (most) vendor files blacklisted

To use this tool, simply run either `uncrustify.sh` for applying the configuration or `check.sh` to check if files have been uncrustified.

Does it make sense? Please feel free to give feedback.

### Issues/PRs references

#4363

### CODE REVIEW

**Comment by jia200x** [comment]
we might need to tune again the cfg file though...

**Comment by gebart** [comment]
Does this have any kind of safety check so that uncrustify does not overwrite local working tree changes?

File: dist/tools/uncrustify/check.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+RIOTBASE=$(git rev-parse --show-toplevel)
+CURDIR=$(cd $(dirname $0) && pwd)
+BLACKLIST=$CURDIR/blacklist.txt
+FILES=$(git ls-tree -r HEAD --name-only --full-tree | grep -E '.*\.(c$|cpp$|h$|hpp$)' | grep -vf $BLACKLIST)
```
Comment (kaspar030) [inline dist/tools/uncrustify/check.sh]: if possible, please re-use dist/tools/ci/changed_files.sh

File: dist/tools/uncrustify/uncrustify.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
Comment (kaspar030) [inline dist/tools/uncrustify/uncrustify.sh]: I'd prefer one script with either "--check" or inplace editing. what do you think?

**Comment by kaspar030** [comment]
> we might need to tune again the cfg file though...

that will be the hard part. ;)

File: dist/tools/uncrustify/check.sh (hunk 1 above)
Comment (jia200x) [reply dist/tools/uncrustify/check.sh]: ah, sure! Didn't know such file existed

File: dist/tools/uncrustify/uncrustify.sh (hunk 2 above)
Comment (jia200x) [reply dist/tools/uncrustify/uncrustify.sh]: yes, sounds reasonable. I will push a fix of this.

**Comment by jia200x** [comment]
> Does this have any kind of safety check so that uncrustify does not overwrite local working tree changes?

Maybe I could add something that prevents the execution if there are changes (e.g output with `git diff`)

File: dist/tools/uncrustify/check.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+for F in $FILES
+do
+	OUT=$(diff $RIOTBASE/$F <(uncrustify -c $CURDIR/uncrustify-riot.cfg -f $RIOTBASE/$F 2>/dev/null))
```
Comment (aabadie) [inline dist/tools/uncrustify/check.sh]: tab ?

File: dist/tools/uncrustify/uncrustify.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+RIOTBASE=$(git rev-parse --show-toplevel)
+CURDIR=$(cd $(dirname $0) && pwd)
+BLACKLIST=$CURDIR/blacklist.txt
+FILES=$(git ls-tree -r HEAD --name-only --full-tree | grep -E '.*\.(c$|cpp$|h$|hpp$)' | grep -vf $BLACKLIST)
+for F in $FILES
+do
+	uncrustify -c $CURDIR/uncrustify-riot.cfg --no-backup $RIOTBASE/$F
```
Comment (aabadie) [inline dist/tools/uncrustify/uncrustify.sh]: Indentation should be 4 spaces (here and above)

**Comment by jia200x** [comment]
@aabadie solved

**Comment by bergzand** [comment]
@jia200x Could you please run your shell scripts through [shellcheck](https://www.shellcheck.net/)?
It checks for simple issues such as missing quotes around variables (can cause issues with spaces in paths).

I can enumerate the issues as a review for you if you have trouble using shellcheck.

**Comment by jia200x** [comment]
@bergzand cool! Ran OK.  I'm fixing those issues.

**Comment by jia200x** [comment]
@bergzand done. Thanks for the tip!

File: dist/tools/uncrustify/uncrustify.sh (hunk 2 above)
Comment (jia200x) [reply dist/tools/uncrustify/uncrustify.sh]: addressed!

File: dist/tools/uncrustify/uncrustify.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+RIOTBASE=$(git rev-parse --show-toplevel)
+CURDIR=$(cd "$(dirname "$0")" && pwd)
+BLACKLIST=$CURDIR/blacklist.txt
+
+. "$RIOTBASE"/dist/tools/ci/changed_files.sh
+FILES=$(changed_files | grep -vf "$BLACKLIST")
```
Comment (kaspar030) [inline dist/tools/uncrustify/uncrustify.sh]: changed files allows it's own blacklist. Sth like this should be equivalent:

```FILES=$(EXCLUDE="(.*/vendor/.*|.*/core/.*)" changed_files)```

Comment (jia200x) [reply dist/tools/uncrustify/uncrustify.sh]: I proposed to provide a separate blacklist.txt file because I think it's easier to add/remove entries from there. Since there might be several blacklisted entries, I thought it was a good idea. What do you think?

Comment (kaspar030) [reply dist/tools/uncrustify/uncrustify.sh]: > What do you think?

Ok! Then I'd suggest removing "vendor", as that should be filtered by changed_files' defaults.
Anyone reading this, next one who needs a blacklist file needs to add generic support for that to changed_files.sh! ;)

Comment (jia200x) [reply dist/tools/uncrustify/uncrustify.sh]: Ok, perfect! ;)

**Comment by jia200x** [comment]
@kaspar030 @haukepetersen anything else to blackilist besides core files?

**Comment by jia200x** [comment]
(If this get merged, the uncrustify process should come in a further PR)

**Comment by kaspar030** [comment]
> @kaspar030 @haukepetersen anything else to blackilist besides core files?

Well, as is uncrustify would change ~24k lines, with many of them unwanted changes. That IMO shows that the config file needs some tuning... I think it'll take some time util we can actually *force* uncrustify, even for new files. Thus whether core is blacklisted or not is pretty much irrelevant...

**Comment by jia200x** [comment]
> Well, as is uncrustify would change ~24k lines, with many of them unwanted changes. That IMO shows that the config file needs some tuning... I think it'll take some time util we can actually force uncrustify, even for new files. Thus whether core is blacklisted or not is pretty much irrelevant...

Could you point to some unwanted changes? ;) I can do my best to address them.

**Comment by jia200x** [comment]
@kaspar030 well, assuming files in upstream are in good shape, I wil try to find the config that minimizes the number of changed files

**Comment by kaspar030** [comment]
This is from the first diff uncrustify.sh does on this PR's branch:

```
         .name = "Button 1",
-        .pin  = BTN0_PIN,
+        .pin = BTN0_PIN,
         .mode = BTN0_MODE,
         .flags = SAUL_GPIO_INVERTED,
     }
```

The alignment is removed here.

**Comment by jia200x** [comment]
> The alignment is removed here.
I see.. I will play a little bit with the cfg file

**Comment by tcschmidt** [comment]
This tool seems a proper way to automate the burden of code style review. Can we advance to make this production ready for
 * sanitizing old code
 * adding this to the automated test system so that new PRs authors get automatic feedback prior to review?

**Comment by haukepetersen** [comment]
good luck :-) All the previous attempts to make this work over the last years have failed - but it would be nice to get somewhere this time!

I'd say, if applying uncrustify to our codebase yields somwhere < 1000 changes, we are on a good path...

**Comment by tcschmidt** [comment]
I'm sure @jia200x and others will have energy to make this work: efforts most likely will be honored by a huge burden of labor taken off reviewers shoulders.
Btw @haukepetersen : How is the status of your attempt for structuring review work? You started out with a Wiki page a while ago ?

**Comment by jia200x** [comment]
Check [this gist](https://gist.github.com/jia200x/5d9d2b61c036814efb957a4e3c367ff4) to see the how the whole RIOT repo (without core files) get's uncrustified.

**Comment by kaspar030** [comment]
As discussed offline, let's go with an iterative approach here.

One way would be:

- add the result of "uncrustify -c /dev/null --update-config > uncrustify-force.cfg" somewhere. Best make a commit of it's own, stating the used uncrustify version

- try to disable as much as possible in that config, set defaults for the rest, in order to get the minimal amount of lines when applied to the code base

- those lines left, either accept the changes, or exclude from uncrustify (using uncrustify's macros)

- merge this PR, with the enforcment enabled

Then, successively, add more options from uncrustify-riot.cfg. Probably best would be one per PR to make the discussions both manageable and parallelizable.

**Comment by tcschmidt** [comment]
@jia200x ready for the first deployment step as outlined by @kaspar030 ?

**Comment by jia200x** [comment]
@tcschmidt @kaspar030 I already pushed some updated with a minimum uncrustify-force.cfg file. The uncrustify-riot.cfg file is untouched.

Check [this gist](https://gist.github.com/jia200x/5d9d2b61c036814efb957a4e3c367ff4). IMO 80~90% of the changes are ok (or reasonably OK). There are a few corner cases which could be disabled with Uncrustify directives `/* INDENT-ON */` and `/* INDENT-OFF */`. E.g

```
  * @name   Timer configuration
@@ -127,9 +127,9 @@ static const pwm_conf_t pwm_config[] = {
     {
         .dev      = TIM2,
         .rcc_mask = RCC_APB1ENR_TIM2EN,
-        .chan     = { { .pin = GPIO_PIN(PORT_B, 3)  /* D3 */, .cc_chan = 1 },
+        .chan     = { { .pin = GPIO_PIN(PORT_B, 3) /* D3 */, .cc_chan = 1 },
                       { .pin = GPIO_PIN(PORT_B, 10) /* D6 */, .cc_chan = 2 },
-                      { .pin = GPIO_PIN(PORT_B, 11)         , .cc_chan = 3 },
+                      { .pin = GPIO_PIN(PORT_B, 11), .cc_chan = 3 },
                       { .pin = GPIO_UNDEF,                    .cc_chan = 0 } },
         .af       = GPIO_AF2,
         .bus      = APB1
```

**Comment by danpetry** [comment]
Hey, we found a case where uncrustify is wrong:
 https://github.com/RIOT-OS/RIOT/pull/9346
see discussion about halfway down the page.

**Comment by danpetry** [comment]
Ah! great thanks.

**Comment by jia200x** [comment]
about the comment on #9346 , that's an example that can be fixed with `/* INDENT-ON */` and `/* INDENT-OFF */`. Any objections with the current config file? ;)

**Comment by danpetry** [comment]
Fine with me

**Comment by jia200x** [comment]
@kaspar030 are you OK with the approach?

**Comment by jia200x** [comment]
I've found a much better solution called [Vera++](https://bitbucket.org/verateam/vera/overview)). Contrary to Uncrustify, Vera++ is a code validator instead of a code beautifier. It has a [bunch of predefined rules](https://bitbucket.org/verateam/vera/wiki/Rules) that can be extended with TCL, Lua or Python. I will open another PR and close this one afterwards

**Comment by kaspar030** [comment]
> I've found a much better solution called Vera++

Why better?

Uncrustify, with it's deficiencies, has some very valuable properties:

- it is available in most distro repositories
- it needs only one configuration file
- we have some experience with it

Maybe vera++ can be, with TCL, Lua or Pyton scripts, be made to nag about and transform code as we want it to. But to me it seems *way* too complicated [1] to set up and use. We want our contributors to spend the least amount of time for code formatting as possible.

[1] at least more than ```apt-get install uncrustify```, ```uncrustify -c riot/uncrustify.conf *.c```

**Comment by jia200x** [comment]
hi @kaspar030 

> I've found a much better solution called Vera++

... for checking coding convention from a CI perspective. Uncrustify can/should be still used.

> Maybe vera++ can be, with TCL, Lua or Pyton scripts, be made to nag about and transform code as we want it to. But to me it seems way too complicated [1] to set up and use. We want our contributors to spend the least amount of time for code formatting as possible.

The problem with Uncrustify (at least on CI level) is the fact it's not a code validator but a code beautifier (contrary to Vera++ which is a code validator and transformator). With Uncrustify the only way to check if a file follows RIOT conventions is to compare the uncrustified file with the original. If they don't match, "the file doesn't follow the coding conventions".

I see some problems with this:
- There's no way to know what's the cause of the error (line > 100, extra empty line, etc).
- [Even with the simplest configuration](https://github.com/RIOT-OS/RIOT/pull/8519#issuecomment-396660228), Uncrustify changes most files and tends to give false positives.
- There are some cases where Uncrustify can handle our coding convention:
```
for(...); //This is my comment
```
goes to
```
for(...) {
    //This is my comment
}
```
instead of
```
for (...) {} //This is my comment
```
I still think Uncrustify is easier to use for beautifying the code. Although it's possible with Vera++, this is not the main focus of the tool (it's better for tasks like "change prefix of all macros that match a criteria" or "change the path of header files in #include directives")

IMO Uncrustify can be a developer tool for beautifying the code and Vera++ a tool running in the CI that checks coding convention.

What do you think?

**Comment by kaspar030** [comment]
> There's no way to know what's the cause of the error (line > 100, extra empty line, etc).

Show diff?

> Even with the simplest configuration, Uncrustify changes most files and tends to give false positives.

Yes, that sucks. But it is a problem *we* can solve (by accepting the changes or wrapping in IDENT-OFF/ON).

> There are some cases where Uncrustify can handle our coding convention:

Yup. But I think we could agree to adapt the coding conventions in those cases.

See, I think the one goal we should aim at is to have automatic, consistent code formatting. @tcschmidt was completely right suggesting we should automate fixing *all* style issues in order to cut review time. For that we need a tool with uncrustify's simplicity and availability.

(Vera++ might be useful to go beyond that, e.g., semantic stuff that goes beyond code beautification).

**Comment by jia200x** [comment]
> See, I think the one goal we should aim at is to have automatic, consistent code formatting. @tcschmidt was completely right suggesting we should automate fixing all style issues in order to cut review time. For that we need a tool with uncrustify's simplicity and availability.

My point is, maintainers will still need to do a lot of hand work for tagging all unrelated changes and find agreements in corner cases. If errors are spot directly in the code (without corner cases, where Vera++ seems to be a closer than Uncrustify), maintainers shouldn't spend time trying to find those kind of errors. For people submitting a PR is quite easy to fix style errors if they are known.

**Comment by jia200x** [comment]
may I propose to test both Uncrustify and Vera++ in the CI for a couple of weeks and see which one behaves better for reducing review time?

**Comment by tcschmidt** [comment]
Makes sense to me, but it will be important to have both tools configured as suitable as possible. Otherwise we will rather judge on the quality of configuration ;)

**Comment by jia200x** [comment]
> Makes sense to me, but it will be important to have both tools configured as suitable as possible. Otherwise we will rather judge on the quality of configuration ;)

Sure, we will.

Then can we merge this PR? [This](https://gist.github.com/jia200x/5d9d2b61c036814efb957a4e3c367ff4) is the the whole RIOT repository with the proposed uncrustify-force.cfg

**Comment by jia200x** [comment]
As discussed offline some time ago after the last comment (https://github.com/RIOT-OS/RIOT/pull/8519#issuecomment-414953616), we will opt for Vera++ first.

I will close this PR. Please re-open if needed.

**Comment by kaspar030** [comment]
I would like to revive this.
IMO, what I've seen from Vera++, it is not the right tool to enforce a certain code beautifying. uncrustify is much easier to set up for our users.

As far as I can see, the reasons we're not enforcing a certain uncrustify configuration are:

1. we cannot get consensus for a certain style
2. uncrustify cannot be configured to enforce exactly the style we would like it to

To be honest, I think we should make uncrustify a participant of the discussion and give it 100 times the weight, as in, "guys *this* is what can be done".

Things like this:

>    There are some cases where Uncrustify can handle our coding convention:

```
for(...); //This is my comment

goes to

for(...) {
    //This is my comment
}

instead of

for (...) {} //This is my comment
```

Do they matter that much? Personally I do not care anymore, if in turn I do not have to *ever* deal with formatting anymore. (Unless I want to do ASCII art, but then I can do INDENT-OFF.)

I do see that currently we'd have thousands of lines that would be changed, many of which we don't want to be changed.

Why don't we go with a whitelist first? That way we can slowly adapt the codebase.

I suggest the following:

1. from this PR, remove the "uncrustify-force" config, just use "uncrustify-riot.conf" from the root
2. in addition to the blacklist file, add a whitelist. Start small (e.g., just with core/).
3. get consensus on reduced maintanance / review time over "I prefer this or that style", make "uncrustify-riot.cfg" authorative
4. add a message to the coding convention for that
5. add a wiki page explaining why and how uncrustify

@jia200x If you don't mind I'll add commits for the whitelist.

**Comment by kaspar030** [comment]
> @jia200x If you don't mind

Do you? I have a rebase and addition of whitelist in the queue.

**Comment by jia200x** [comment]
> Do you? I have a rebase and addition of whitelist in the queue.

No problem ;)

**Comment by jia200x** [comment]
Hi @kaspar030 

I completely agree with your comments and agree we should adapt our coding convention a bit to make it easier for us. I also think the corner cases don't matter that much and I prefer to add them to the coding convention instead of  /* INDENT-ON */ and /* INDENT-OFF */ everywhere.

So up to this point, I agree we should improve our Uncrustify integration and it should be tool ran by RIOT developers.


My whole point of this PR is there are 2 different topics:
1. Code beautifying
2. Code-style checking

(1) Is solved by Uncrustify. (2) Is solved by a code-style checker like Vera++

IMO from the CI we should check the style convention and report **why** there's an error somewhere and **where** is the error. The intention of this PR was to add a CI helper, that's why I closed it.

This IMO can't be done properly with Uncrustify.

Please check [this gist](https://gist.github.com/jia200x/f17c6c9316f4f35bf151fab0b112fd7a) with Uncrustify and Vera++ results fro style checking.
I took as an example the xtimer_core.c file (original.c) and ran uncrustify with our current RIOT config (uncrustify.c).

I ran Vera++ in the original file (output at vera_original.txt) and in the uncrustified version (vera_uncrustify.txt).
I also got diff between original.c and uncrustify.c (diff.txt).

Some conclusions:
1. Uncrustify diff shows [some unrelated changes like this](https://gist.github.com/jia200x/f17c6c9316f4f35bf151fab0b112fd7a#file-diff-txt-L15). I wouldn't put a `/* INDENT-OFF */` there because I think it looks more beautiful, but this line is not violating our standard.

2. It's not possible to know **why** there was an issue. Check e.g this snippet:
```
31c531
<         while (reference < _xtimer_lltimer_now()) {}
---
> while( reference < _xtimer_lltimer_now()){}
```

Just by looking the output, what's the error?

Vera++'s original.c outputs:
```
/tmp/original.c:531: keyword 'while' not followed by a single space
```

It's possible to know **where** is the issue by diff output, but if it's unsure if it's an issue then there would be tons of false positives.
And also, isn't Vera++ output exactly what we would like to have in the CI?

3. After running Uncrustify, Vera++ output improves. So it can still increase the chances a PR passes CI style checks.


>- from this PR, remove the "uncrustify-force" config, just use "uncrustify-riot.conf" from the root in addition to the blacklist file, add a whitelist. Start small (e.g., just with core/).
>- get consensus on reduced maintanance / review time over "I prefer this or that style", make "uncrustify-  riot.cfg" authorative
>- add a message to the coding convention for that
> - add a wiki page explaining why and how uncrustify

I agree.

I also propose:
- Adapt this PR to be a helper for the user, not for the CI.
- As discussed previously, enable Vera++ in Travis.

**Comment by jia200x** [comment]
TL;DR:
- Improve Uncrustify support on RIOT and user integration
- Use Vera++ in the CI to check coding style.

In practice, Vera++ would never be run by a developer unless someone wants to run CI checks locally

**Comment by kaspar030** [comment]
>     1. Code beautifying
>     2. Code-style checking

Could you recap where you think they differ?

>  The intention of this PR was to add a CI helper, that's why I closed it.

Yes, that's why I re-opened it. The idea is awesome. I want this to be run on every PR. 

>     * Improve Uncrustify support on RIOT and user integration
>     * Use Vera++ in the CI to check coding style.
> 
> In practice, Vera++ would never be run by a developer unless someone wants to run CI checks locally

To me this defeats the purpose of an automatic code formatting tool, if the developer does not run it.

Please let us converge to a point where we do not have to deal with code formatting anymore.

Last time I looked, Vera++ was far from ```apt-get install vera++; vera++ vera-riot.cfg```. IMO, that ease of use is a requirement.

**Comment by kaspar030** [comment]
> * Adapt this PR to be a helper for the user, not for the CI.

As is, it can do both.

**Comment by jia200x** [comment]
Just to make sure I understood. You plan to transform the code in the CI or just report coding style?

**Comment by kaspar030** [comment]
> Just to make sure I understood. You plan to transform the code in the CI or just report coding style?

Ah, CI just checks. User transforms.

**Comment by jia200x** [comment]
> Well, I mean, CI runs uncrustify in "--check" mode. The user might then want to run uncrustify actually applying the changes.

NB the only thing we can report is wether uncrustify was ran or not (diff != empty). We cannot report what were the errors. 

> To me this defeats the purpose of an automatic code formatting tool, if the developer does not run it.

That's the point, Vera++ is not a code formatting/beautifier tool. It's just a rule-based code style checker.
[It's straightforward to write RIOT convention with Vera++ rules](https://bitbucket.org/verateam/vera/wiki/Rules) and can detect if a rule is violated. Uncrustify parses the C code and rebuilds it based on rules. All information about broken rules is lost.

> Please let us converge to a point where we do not have to deal with code formatting anymore.

If the CI runs coding style checks, we don't have to care about coding style anymore.

E.g these 2 cases:
- Uncrustify running on CI: We can either report:
```

217,222c217,222
<     * Backoff condition above ensures that 'target - XTIMER_OVERHEAD` is later
<     * than 'now', also for values when now will overflow and the value of target
<     * is smaller then now.
<     * If `target < XTIMER_OVERHEAD` the new target will be at the end of this
<     * 32bit period, as `target - XTIMER_OVERHEAD` is a big number instead of a
<     * small at the beginning of the next period. */
---
>      * Backoff condition above ensures that 'target - XTIMER_OVERHEAD` is later
>      * than 'now', also for values when now will overflow and the value of target
>      * is smaller then now.
>      * If `target < XTIMER_OVERHEAD` the new target will be at the end of this
>      * 32bit period, as `target - XTIMER_OVERHEAD` is a big number instead of a
>      * small at the beginning of the next period. */
269c269
<                || (((*list_head)->long_target == timer->long_target) && ((*list_head)->target <= timer->target)))) {
---
>            || (((*list_head)->long_target == timer->long_target) && ((*list_head)->target <= timer->target)))) {
531c531
<         while (reference < _xtimer_lltimer_now()) {}
---
> while( reference < _xtimer_lltimer_now()){}
```

or something like 

```
uncrustify --check failed! Please run Uncrustify
```

- Vera++ running on CI. It would report:

```
/tmp/original.c:269: line is longer than 100 characters
/tmp/original.c:404: line is longer than 100 characters
/tmp/original.c:502: line is longer than 100 characters
/tmp/original.c:531: keyword 'while' not followed by a single space
```

In both cases, **the user would need to run Uncrustify anyway**.

But we have a more detailed report with Vera++, it won't give false positives if we change its rules and the most important thing IMO: we can already run Vera++ _now_  without adding `/* INDENT-OFF */` tags and would check the whole repo as it is.
This doesn't mean we shouldn't gradually add `/* INDENT-OFF */` tags, but at least we can already enforce code style for all RIOT files ASAP with Vera++.

In fact, that's the way how I've been checking coding style on PRs.

**Comment by miri64** [comment]
I agree with @jia200x that Vera++ is the better tool to detect code style errors from a CI perspective and with @kaspar030 that `uncrustify` is the better user tool. IMHO we should also run `uncrustify` in addition to Vera++ on the CI. Vera++ to provide a comprehensive list of quality defects and `uncrustify` to check if its output is still in line with Vera++'s assessment.

**Comment by miri64** [comment]
> (Vera++ happens to not be packaged for my system (current Arch), and it seems like it needs more than one file for its configuration, so my personal preference is strongly towards uncrustify for simplicity's sake.)

[Untrue](https://aur.archlinux.org/packages/vera%2B%2B/) (or are you not using AUR out of principle?)

**Comment by kaspar030** [comment]
> [Untrue](https://aur.archlinux.org/packages/vera%2B%2B/) (or are you not using AUR out of principle?)

You caught me. :) For the argument, obviously only distribution-supported repos count. :grin:

**Comment by kaspar030** [comment]
> In both cases, **the user would need to run Uncrustify anyway**.

Hey, if we can agree that by doing so a user fulfills all formatting requirements, we have converged somewhere.

**Comment by jia200x** [comment]
> But it would output a lot of warnings, right?

Not really. In fact, rules are quite clear. E.g
```
L001 No trailing whitespace
L002 Don't use tab characters
L003 No leading and no trailing empty lines
L004 Line cannot be too long
L005 There should not be too many consecutive empty lines
L006 Source file should not be too long
```

This rules are not ambiguous at all. They don't produce false positives.

> Well, I'd like to get you (and every human) out of the loop. 

Yes, I just meant I was able to report errors using Vera++ output. So the CI could eventually do the same.

> Even looking at warnings regarding whitespaces is a waste of time in 99% of the cases.

Just one space is enough for rejecting the code style. So I don't see this as a problem.

> in order to comply with our formatting run $tool with $config

We can do that with Uncrustify. I think that's not my point of discussion.

> but if it is not the right tool to be used for fully automatic reformatting on the users end, why bother?

As I said before, I agree with this. I wouldn't expect a developer to use Vera++. You could also run `cppcheck` as a developer but it's intended for the CI

> I don't care which tool does it in the end, I just don't want to be bothered but still have consistent code looks.
(Vera++ happens to not be packaged for my system (current Arch), and it seems like it needs more than one file for its configuration, so my personal preference is strongly towards uncrustify for simplicity's sake.)

I really get this one. And I think we should use Uncrustify for the users.

> Maybe. But a very important aspect of reducing reviewing overhead is to have $tool that is authoritative on all things formatting. No discussion. No second opinion. Let's have whitespace questions be answered fully automatic.

I can confirm Vera++ does EXACTLY what the rules say. There's no second opinion to "Extra empty linespace" or "Line too long". You can check the rules [here](https://bitbucket.org/verateam/vera/wiki/Rules). There are not so many, and they are already incorporated in https://github.com/RIOT-OS/RIOT/pull/9778.

> (And honestly, I think we should avoid wasting time getting the tools in sync, which will also be a waste of time, if a simple tool with agreement on one configuration solves the problem, IMO...)

Same as before. There's not too much to get in sync there. Unless we decide to remove the 80-100 lines restrictions or trailing whitespaces, we shouldn't have issues with this.

> Hey, if we can agree that by doing so a user fulfills all formatting requirements, we have converged somewhere.

I think we have converged to the usage of Uncrustify on the developer/user side.
IMO the point of the discussion has to do with the usage of Vera++ and/or Uncrustify in the CI

**Comment by miri64** [comment]
Let's do this humming style add thumbs up to what you prefer.

**Comment by miri64** [comment]
Only Vera++ (on CI and for users)

**Comment by miri64** [comment]
Vera++ and Uncrustify (on CI and users can just use Uncrustify)

**Comment by miri64** [comment]
Vera++ on CI, Uncrustify for users.

**Comment by danpetry** [comment]
Vera++ and uncrustify are AFAIU different tools which provide different checks. And uncrustify can't test for adherence to the [Coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions) whereas Vera++ can.

+1 for applying both on the CI. Don't quite understand "developers can just use uncrustify", though? I guess both config files will be available so they could use both?

**Comment by danpetry** [comment]
Edit: upon closer inspection, uncrustify can test some coding conventions, vera++ can test others...

**Comment by miri64** [comment]
> Don't quite understand "developers can just use uncrustify", though? I guess both config files will be available so they could use both?

The problem is that Vera++ still requires manual editing as far as I remember. So if Vera++ and uncrustify are synced on their "opinion" of style, uncrustify is the better tool for a developer to clean up their code. That was what I meant with that.

**Comment by kaspar030** [comment]
> Not really. In fact, rules are quite clear. E.g
> 
> ```
> L001 No trailing whitespace
> L002 Don't use tab characters
> L003 No leading and no trailing empty lines
> L004 Line cannot be too long
> L005 There should not be too many consecutive empty lines
> L006 Source file should not be too long
> ```
> 
> This rules are not ambiguous at all. They don't produce false positives.

Ok, now I see. Vera++ only checks for compliance to those rules. Most code complies to that already, so arming it now would be painless. More rules can be added later.

Obviously Vera++ and uncrustify would need to agree on their style descisions. Otherwise users might run uncrustify over code that Vera++ likes, just to have Vera++ complain afterwards, or the other way around.

Looking at the rules, there are already some that uncrustify doesn't care about (many consecutive empty lines, source file too long), but which make sense for the CI.

It would make total sense to use Vera++ for the checks uncrustify cannot do. Still I believe that having the (much simpler) uncrustify as authoritative baseline formatting would make devs and reviewers life a little easier.
As long as the tools check different (or guaranteed to be the same) stuff, I'm happy to use both.

**Comment by jia200x** [comment]
> Ok, now I see. Vera++ only checks for compliance to those rules. Most code complies to that already, so arming it now would be painless. More rules can be added later.

Yes, exactly :)

> Obviously Vera++ and uncrustify would need to agree on their style descisions. Otherwise users might run uncrustify over code that Vera++ likes, just to have Vera++ complain afterwards, or the other way around.

Yes, that is a must. So far, I haven't detected any conflicts. Basically because Vera++ doesn't have too much to do with alignments, rearrangements, etc but more with stuff like `T007 Semicolons should not be isolated by spaces or comments from the rest of the code` or `T009 Comma should not be preceded by whitespace, but should be followed by one`.

> Still I believe that having the (much simpler) uncrustify as authoritative baseline formatting would make devs and reviewers life a little easier.
As long as the tools check different (or guaranteed to be the same) stuff, I'm happy to use both.

Yes, sure. Re-checking Vera++ rules, I think it will be hard to produce Uncrustify output that might conflict with Vera++. So, it would be safe to use Uncrustify as authoritative baseline formatting on the dev side.

**Comment by kaspar030** [comment]
I still want this!

**Comment by jia200x** [comment]
@kaspar030 are you still interested in this PR?

**Comment by kaspar030** [comment]
> @kaspar030 are you still interested in this PR?

yes! I want this to enforce uncrustify once #10867 is in.

**Comment by kaspar030** [comment]
https://github.com/RIOT-OS/RIOT/pull/10867 is merged.
Could you rebase, move the check to `dist/tools/ci/static_tests.sh` and add `core/*.c` and `core/include/*.h`?

**Comment by jia200x** [comment]
> Could you rebase, move the check to dist/tools/ci/static_tests.sh and add core/*.c and core/include/*.h?

Sure!

**Comment by kaspar030** [comment]
*ping* @jia200x, I have review time tomorrow!

**Comment by jia200x** [comment]
done!

**Comment by jia200x** [comment]
Side note: In another open source project I'm working as a volunteer we are using a linter and a code beautifier. Both are for python code (black and pylint) but the experience has been quite nice.

The CI complains if the developer didn't run "black" (beautifier) but also if the linter detect issues (pylint). So I think this can live in harmony with Vera++

**Comment by kaspar030** [comment]
LGTM, please squash!

**Comment by jia200x** [comment]
squashed!

**Comment by kaspar030** [comment]
CI passed on compilation but failed on (still set) "needs squashing". re-triggering.

**Comment by kaspar030** [comment]
@jia200x I found an issue: the whitelist contains paths, but should contain regexps. Also, each line should match a whole path, not a substring.
So as is the whitelist entry only matches by chance.
And when the regexp is fixed to e.g., `core/.*\.c`, it would also match `foo/.../tests-core/foo.c`.

This patch should fix the issue (adds "-x" for line based patterns to the filter, and fixes the core regexps):

https://gist.github.com/e28bea3fd46a16651ea955f5abf0b2b7

**Comment by jia200x** [comment]
fixed, and squashed directly. Thanks for the patch!

**Comment by jia200x** [comment]
does this require Murdock? Can't we add the `CI: skip compile tests` label?
